File: frontend/src/features/payables/services/payableService.ts
import { INITIAL_PAYABLE_CATEGORIES } from "../data/initialPayableCategories";
import {
  getEffectiveStatus,
  getMonthKey,
  isDueWithinDays,
  matchesMonthFilter,
  matchesStatusFilter,
} from "../utils/payableStatus";
import { calculateNextDueDate, getRecurrenceLabel } from "../utils/recurrence";
import type {
  Payable,
  PayableActionResult,
  PayableFormInput,
  PayableManualStatus,
  PayableRecurrence,
  PayableStatusFilter,
  PayableSummary,
} from "../types";

export function validatePayableInput(
  input: PayableFormInput,
  supplierExists: (supplierId: string) => boolean,
): PayableActionResult {
  if (!input.description.trim()) {
    return { ok: false, error: "Informe uma descrição." };
  }

  if (!Number.isFinite(input.amount) || input.amount <= 0) {
    return { ok: false, error: "Informe um valor válido maior que zero." };
  }

  if (!/^\d{4}-\d{2}-\d{2}$/.test(input.dueDate)) {
    return { ok: false, error: "Informe uma data de vencimento válida." };
  }

  const categoryExists = INITIAL_PAYABLE_CATEGORIES.some(
    (category) => category.id === input.categoryId,
  );
  if (!categoryExists) {
    return { ok: false, error: "Selecione uma categoria válida." };
  }

  if (input.supplierId && !supplierExists(input.supplierId)) {
    return { ok: false, error: "Selecione um fornecedor válido." };
  }

  if (input.status === "paid") {
    if (!input.paidAt) {
      return { ok: false, error: "Informe a data do pagamento." };
    }

    const paidAmount = input.paidAmount ?? input.amount;
    if (!Number.isFinite(paidAmount) || paidAmount <= 0) {
      return { ok: false, error: "Informe um valor pago válido." };
    }
  }

  return { ok: true };
}

export function buildPayableFromInput(
  input: PayableFormInput,
  id: string,
  createdAt: string,
  current?: Payable,
): Payable {
  const status: PayableManualStatus =
    input.status ?? (current?.status === "paid" ? "paid" : "pending");
  const base = {
    id,
    categoryId: input.categoryId,
    description: input.description.trim(),
    supplierId: input.supplierId || undefined,
    amount: input.amount,
    dueDate: input.dueDate,
    recurrence: input.recurrence,
    notes: input.notes?.trim() || undefined,
    createdAt: current?.createdAt ?? createdAt,
  };

  if (status === "paid") {
    return {
      ...base,
      status: "paid",
      paidAt: input.paidAt ?? current?.paidAt ?? new Date().toISOString(),
      paidAmount: input.paidAmount ?? current?.paidAmount ?? input.amount,
    };
  }

  return {
    ...base,
    status: "pending",
    paidAt: undefined,
    paidAmount: undefined,
  };
}

export function buildNextRecurringPayable(paid: Payable, newId: string): Payable | null {
  const nextDueDate = calculateNextDueDate(paid.dueDate, paid.recurrence);
  if (!nextDueDate) {
    return null;
  }

  return {
    id: newId,
    categoryId: paid.categoryId,
    description: paid.description,
    supplierId: paid.supplierId,
    amount: paid.amount,
    dueDate: nextDueDate,
    recurrence: paid.recurrence,
    status: "pending",
    notes: paid.notes,
    createdAt: new Date().toISOString(),
  };
}

export function applyPaidStatus(
  payable: Payable,
  paidAt: string,
  paidAmount: number,
): Payable {
  return {
    ...payable,
    status: "paid",
    paidAt,
    paidAmount,
  };
}

export function applyPendingStatus(payable: Payable): Payable {
  return {
    ...payable,
    status: "pending",
    paidAt: undefined,
    paidAmount: undefined,
  };
}

export function filterPayables(
  payables: Payable[],
  statusFilter: PayableStatusFilter,
  monthFilter: string,
  searchQuery: string,
  resolveSupplierName: (supplierId: string | undefined) => string,
  referenceDate = new Date(),
): Payable[] {
  const normalizedQuery = searchQuery.trim().toLowerCase();

  return payables.filter((payable) => {
    if (!matchesStatusFilter(payable, statusFilter, referenceDate)) {
      return false;
    }

    if (!matchesMonthFilter(payable, monthFilter, referenceDate)) {
      return false;
    }

    if (!normalizedQuery) {
      return true;
    }

    const categoryName = getCategoryName(payable.categoryId);
    const supplierName = resolveSupplierName(payable.supplierId);
    const recurrenceLabel = getRecurrenceLabel(payable.recurrence);

    const haystack = [payable.description, supplierName, categoryName, recurrenceLabel]
      .join(" ")
      .toLowerCase();

    return haystack.includes(normalizedQuery);
  });
}

export function computePayableSummary(
  payables: Payable[],
  referenceDate = new Date(),
): PayableSummary {
  const currentMonthKey = getMonthKey(referenceDate);

  let dueSoonCount = 0;
  let dueSoonTotal = 0;
  let overdueCount = 0;
  let overdueTotal = 0;
  let paidThisMonthCount = 0;
  let paidThisMonthTotal = 0;
  let pendingThisMonthTotal = 0;

  for (const payable of payables) {
    const effectiveStatus = getEffectiveStatus(payable, referenceDate);

    if (isDueWithinDays(payable, 7, referenceDate)) {
      dueSoonCount += 1;
      dueSoonTotal += payable.amount;
    }

    if (effectiveStatus === "overdue") {
      overdueCount += 1;
      overdueTotal += payable.amount;
    }

    if (payable.status === "paid" && payable.paidAt) {
      const paidMonthKey = getMonthKey(new Date(payable.paidAt));
      if (paidMonthKey === currentMonthKey) {
        paidThisMonthCount += 1;
        paidThisMonthTotal += payable.paidAmount ?? payable.amount;
      }
    }

    const dueDate = payable.dueDate.slice(0, 7);
    if (
      dueDate === currentMonthKey &&
      (effectiveStatus === "pending" || effectiveStatus === "overdue")
    ) {
      pendingThisMonthTotal += payable.amount;
    }
  }

  return {
    dueSoonCount,
    dueSoonTotal,
    overdueCount,
    overdueTotal,
    paidThisMonthCount,
    paidThisMonthTotal,
    pendingThisMonthTotal,
  };
}

export function getCategoryName(categoryId: string): string {
  return (
    INITIAL_PAYABLE_CATEGORIES.find((category) => category.id === categoryId)?.name ?? "Outros"
  );
}

export { INITIAL_PAYABLE_CATEGORIES, getRecurrenceLabel };
export type { PayableRecurrence };

File: backend/app/services/payable_service.py
import calendar
import uuid
from datetime import date, datetime

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.models.enums import PayableRecurrence, PayableStatus
from app.models.payable import Payable, PayableCategory
from app.models.supplier import Supplier
from app.services.mappers import utc_now


def _add_months(source: date, months: int) -> date:
    month_index = source.month - 1 + months
    year = source.year + month_index // 12
    month = month_index % 12 + 1
    day = min(source.day, calendar.monthrange(year, month)[1])
    return date(year, month, day)


def calculate_next_due_date(due_date: date, recurrence: str) -> date | None:
    if recurrence == PayableRecurrence.NONE:
        return None

    if recurrence == PayableRecurrence.MONTHLY:
        return _add_months(due_date, 1)
    if recurrence == PayableRecurrence.QUARTERLY:
        return _add_months(due_date, 3)
    if recurrence == PayableRecurrence.SEMIANNUAL:
        return _add_months(due_date, 6)
    if recurrence == PayableRecurrence.YEARLY:
        return date(due_date.year + 1, due_date.month, due_date.day)
    return None


def get_effective_status(payable: Payable, reference: date | None = None) -> str:
    if payable.status in {PayableStatus.PAID, PayableStatus.CANCELLED}:
        return payable.status

    today = reference or utc_now().date()
    due = payable.due_date if isinstance(payable.due_date, date) else payable.due_date
    return "overdue" if due < today else PayableStatus.PENDING


def validate_payable_input(
    db: Session,
    category_id: str,
    description: str,
    amount: float,
    supplier_id: str | None,
) -> None:
    if not description.strip():
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Informe uma descrição.")

    if amount <= 0:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Informe um valor válido maior que zero.",
        )

    category = db.get(PayableCategory, category_id)
    if not category:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Selecione uma categoria válida.",
        )

    if supplier_id and not db.get(Supplier, supplier_id):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Selecione um fornecedor válido.",
        )


def build_next_recurring_payable(paid: Payable) -> Payable | None:
    due = paid.due_date if isinstance(paid.due_date, date) else paid.due_date
    next_due = calculate_next_due_date(due, paid.recurrence)
    if not next_due:
        return None

    return Payable(
        id=f"payable-{uuid.uuid4()}",
        category_id=paid.category_id,
        description=paid.description,
        supplier_id=paid.supplier_id,
        amount=paid.amount,
        due_date=next_due,
        recurrence=paid.recurrence,
        status=PayableStatus.PENDING,
        notes=paid.notes,
        created_at=utc_now(),
    )


def compute_payable_summary(payables: list[Payable], reference: datetime | None = None) -> dict:
    ref = reference or utc_now()
    today = ref.date()
    current_month_key = f"{today.year:04d}-{today.month:02d}"

    due_soon_count = 0
    due_soon_total = 0.0
    overdue_count = 0
    overdue_total = 0.0
    paid_this_month_count = 0
    paid_this_month_total = 0.0
    pending_this_month_total = 0.0

    for payable in payables:
        effective = get_effective_status(payable, today)

        if effective == PayableStatus.PENDING:
            due = payable.due_date if isinstance(payable.due_date, date) else payable.due_date
            days_until = (due - today).days
            if 0 <= days_until <= 7:
                due_soon_count += 1
                due_soon_total += payable.amount

        if effective == "overdue":
            overdue_count += 1
            overdue_total += payable.amount

        if payable.status == PayableStatus.PAID and payable.paid_at:
            paid_month = f"{payable.paid_at.year:04d}-{payable.paid_at.month:02d}"
            if paid_month == current_month_key:
                paid_this_month_count += 1
                paid_this_month_total += payable.paid_amount or payable.amount

        due = payable.due_date if isinstance(payable.due_date, date) else payable.due_date
        due_month = f"{due.year:04d}-{due.month:02d}"
        if due_month == current_month_key and effective in {PayableStatus.PENDING, "overdue"}:
            pending_this_month_total += payable.amount

    return {
        "dueSoonCount": due_soon_count,
        "dueSoonTotal": due_soon_total,
        "overdueCount": overdue_count,
        "overdueTotal": overdue_total,
        "paidThisMonthCount": paid_this_month_count,
        "paidThisMonthTotal": paid_this_month_total,
        "pendingThisMonthTotal": pending_this_month_total,
    }
